lib/linked_list.rb

```diff
@@ -1,161 +1,268 @@
-
 # Defines a node in the singly linked list
 class Node
   attr_reader :data # allow external entities to read value but not write
-  attr_accessor :next # allow external entities to read or write next node
+  attr_accessor :next, :previous # allow external entities to read or write next node
 
-  def initialize(value, next_node = nil)
+  def initialize(value, next_node = nil, previous_node = nil)
     @data = value
     @next = next_node
+    @previous = previous_node
   end
 end
 
 # Defines the singly linked list
 class LinkedList
-    def initialize
-      @head = nil # keep the head private. Not accessible outside this class
-    end
+  def initialize
+    @head = nil # keep the head private. Not accessible outside this class
+    @tail = nil
+  end
 
-    # method to add a new node with the specific data value in the linked list
-    # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def add_first(value)
-      raise NotImplementedError
+  # method to add a new node with the specific data value in the linked list
+  # insert the new node at the beginning of the linked list
+  # Time Complexity: O(1), because inserting in the front only requires a change of reference.
+  # Space Complexity: O(1), because no new structures are created.
+  def add_first(value)
+    if @head.nil?
+      @head = @tail = Node.new(value)
+    else
+      new_node = Node.new(value, @head)
+      @head.previous = new_node
+      @head = new_node
     end

```

Comment:
Can be simplified a bit.

```suggestion
    @head = Node.new(value, @head)
```